presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,151 @@
+
+        List<RowExpression> rightRandomExpressions = candidateEquiJoinClauses.stream().map(x -> x.getRight()).map(x -> randomizeJoinKey(x, context, "r")).collect(Collectors.toList());
+        List<VariableReferenceExpression> rightRandomVariable = rightRandomExpressions.stream().map(x -> context.getVariableAllocator().newVariable(x, VARIABLE_SUFFIX)).collect(Collectors.toList());
+
+        checkState(leftRandomVariable.size() == rightRandomVariable.size());
+
+        List<JoinNode.EquiJoinClause> joinClauses = IntStream.range(0, leftRandomVariable.size()).boxed().map(index -> new JoinNode.EquiJoinClause(leftRandomVariable.get(index), rightRandomVariable.get(index))).collect(Collectors.toList());
+        joinClauses.addAll(joinNode.getCriteria());
+
+        Assignments.Builder leftAssignments = Assignments.builder();
+        Map<VariableReferenceExpression, RowExpression> leftProjectAssignments = IntStream.range(0, leftRandomVariable.size()).boxed().collect(ImmutableMap.toImmutableMap(i -> leftRandomVariable.get(i), i -> leftRandomExpressions.get(i)));
+        leftAssignments.putAll(leftProjectAssignments);
+        leftAssignments.putAll(joinNode.getLeft().getOutputVariables().stream().collect(ImmutableMap.toImmutableMap(identity(), identity())));
+
+        Assignments.Builder rightAssignments = Assignments.builder();
+        Map<VariableReferenceExpression, RowExpression> rightProjectAssignments = IntStream.range(0, rightRandomVariable.size()).boxed().collect(ImmutableMap.toImmutableMap(i -> rightRandomVariable.get(i), i -> rightRandomExpressions.get(i)));
+        rightAssignments.putAll(rightProjectAssignments);
+        rightAssignments.putAll(joinNode.getRight().getOutputVariables().stream().collect(ImmutableMap.toImmutableMap(identity(), identity())));
+
+        return Result.ofPlanNode(
+                new JoinNode(
+                        joinNode.getSourceLocation(),
+                        joinNode.getId(),
+                        joinNode.getStatsEquivalentPlanNode(),
+                        joinNode.getType(),
+                        new ProjectNode(context.getIdAllocator().getNextId(), joinNode.getLeft(), leftAssignments.build()),
+                        new ProjectNode(context.getIdAllocator().getNextId(), joinNode.getRight(), rightAssignments.build()),
+                        joinClauses,
+                        joinNode.getOutputVariables(),
+                        joinNode.getFilter(),
+                        joinNode.getLeftHashVariable(),
+                        joinNode.getRightHashVariable(),
+                        joinNode.getDistributionType(),
+                        joinNode.getDynamicFilters()));
+    }
+
+    private RowExpression randomizeJoinKey(RowExpression keyExpression, Context context, String prefix)
```

Comment:
Move this up next to the other private helper methods

Reply:
Done